<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Will Landau" />

<meta name="date" content="2015-01-01" />

<title>fbseqStudies package installation and usage guide</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<link href="data:text/css,body%20%7B%0A%20%20background%2Dcolor%3A%20%23fff%3B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20max%2Dwidth%3A%20700px%3B%0A%20%20overflow%3A%20visible%3B%0A%20%20padding%2Dleft%3A%202em%3B%0A%20%20padding%2Dright%3A%202em%3B%0A%20%20font%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20font%2Dsize%3A%2014px%3B%0A%20%20line%2Dheight%3A%201%2E35%3B%0A%7D%0A%0A%23header%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0A%0A%23TOC%20%7B%0A%20%20clear%3A%20both%3B%0A%20%20margin%3A%200%200%2010px%2010px%3B%0A%20%20padding%3A%204px%3B%0A%20%20width%3A%20400px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20border%2Dradius%3A%205px%3B%0A%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20font%2Dsize%3A%2013px%3B%0A%20%20line%2Dheight%3A%201%2E3%3B%0A%7D%0A%20%20%23TOC%20%2Etoctitle%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%20%20font%2Dsize%3A%2015px%3B%0A%20%20%20%20margin%2Dleft%3A%205px%3B%0A%20%20%7D%0A%0A%20%20%23TOC%20ul%20%7B%0A%20%20%20%20padding%2Dleft%3A%2040px%3B%0A%20%20%20%20margin%2Dleft%3A%20%2D1%2E5em%3B%0A%20%20%20%20margin%2Dtop%3A%205px%3B%0A%20%20%20%20margin%2Dbottom%3A%205px%3B%0A%20%20%7D%0A%20%20%23TOC%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dleft%3A%20%2D2em%3B%0A%20%20%7D%0A%20%20%23TOC%20li%20%7B%0A%20%20%20%20line%2Dheight%3A%2016px%3B%0A%20%20%7D%0A%0Atable%20%7B%0A%20%20margin%3A%201em%20auto%3B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dcolor%3A%20%23DDDDDD%3B%0A%20%20border%2Dstyle%3A%20outset%3B%0A%20%20border%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0A%20%20border%2Dwidth%3A%202px%3B%0A%20%20padding%3A%205px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0A%20%20border%2Dwidth%3A%201px%3B%0A%20%20border%2Dstyle%3A%20inset%3B%0A%20%20line%2Dheight%3A%2018px%3B%0A%20%20padding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0A%20%20border%2Dleft%2Dstyle%3A%20none%3B%0A%20%20border%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Ap%20%7B%0A%20%20margin%3A%200%2E5em%200%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20background%2Dcolor%3A%20%23f6f6f6%3B%0A%20%20padding%3A%200%2E25em%200%2E75em%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20border%2Dstyle%3A%20solid%3B%0A%20%20border%3A%20none%3B%0A%20%20border%2Dtop%3A%201px%20solid%20%23777%3B%0A%20%20margin%3A%2028px%200%3B%0A%7D%0A%0Adl%20%7B%0A%20%20margin%2Dleft%3A%200%3B%0A%7D%0A%20%20dl%20dd%20%7B%0A%20%20%20%20margin%2Dbottom%3A%2013px%3B%0A%20%20%20%20margin%2Dleft%3A%2013px%3B%0A%20%20%7D%0A%20%20dl%20dt%20%7B%0A%20%20%20%20font%2Dweight%3A%20bold%3B%0A%20%20%7D%0A%0Aul%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%7D%0A%20%20ul%20li%20%7B%0A%20%20%20%20list%2Dstyle%3A%20circle%20outside%3B%0A%20%20%7D%0A%20%20ul%20ul%20%7B%0A%20%20%20%20margin%2Dbottom%3A%200%3B%0A%20%20%7D%0A%0Apre%2C%20code%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20color%3A%20%23333%3B%0A%7D%0Apre%20%7B%0A%20%20white%2Dspace%3A%20pre%2Dwrap%3B%20%20%20%20%2F%2A%20Wrap%20long%20lines%20%2A%2F%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20margin%3A%205px%200px%2010px%200px%3B%0A%20%20padding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0A%20%20background%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0A%0Acode%20%7B%0A%20%20font%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0A%20%20font%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0A%20%20padding%3A%202px%200px%3B%0A%7D%0A%0Adiv%2Efigure%20%7B%0A%20%20text%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0A%20%20background%2Dcolor%3A%20%23FFFFFF%3B%0A%20%20padding%3A%202px%3B%0A%20%20border%3A%201px%20solid%20%23DDDDDD%3B%0A%20%20border%2Dradius%3A%203px%3B%0A%20%20border%3A%201px%20solid%20%23CCCCCC%3B%0A%20%20margin%3A%200%205px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20margin%2Dtop%3A%200%3B%0A%20%20font%2Dsize%3A%2035px%3B%0A%20%20line%2Dheight%3A%2040px%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20border%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20padding%2Dbottom%3A%202px%3B%0A%20%20font%2Dsize%3A%20145%25%3B%0A%7D%0A%0Ah3%20%7B%0A%20%20border%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0A%20%20padding%2Dtop%3A%2010px%3B%0A%20%20font%2Dsize%3A%20120%25%3B%0A%7D%0A%0Ah4%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0A%20%20margin%2Dleft%3A%208px%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Ah5%2C%20h6%20%7B%0A%20%20border%2Dbottom%3A%201px%20solid%20%23ccc%3B%0A%20%20font%2Dsize%3A%20105%25%3B%0A%7D%0A%0Aa%20%7B%0A%20%20color%3A%20%230033dd%3B%0A%20%20text%2Ddecoration%3A%20none%3B%0A%7D%0A%20%20a%3Ahover%20%7B%0A%20%20%20%20color%3A%20%236666ff%3B%20%7D%0A%20%20a%3Avisited%20%7B%0A%20%20%20%20color%3A%20%23800080%3B%20%7D%0A%20%20a%3Avisited%3Ahover%20%7B%0A%20%20%20%20color%3A%20%23BB00BB%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%20%20a%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0A%20%20%20%20text%2Ddecoration%3A%20underline%3B%20%7D%0A%0A%2F%2A%20Class%20described%20in%20https%3A%2F%2Fbenjeffrey%2Ecom%2Fposts%2Fpandoc%2Dsyntax%2Dhighlighting%2Dcss%0A%20%20%20Colours%20from%20https%3A%2F%2Fgist%2Egithub%2Ecom%2Frobsimmons%2F1172277%20%2A%2F%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Keyword%20%2A%2F%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%2F%2A%20DataType%20%2A%2F%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%2F%2A%20DecVal%20%28decimal%20values%29%20%2A%2F%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20BaseN%20%2A%2F%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Float%20%2A%2F%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20Char%20%2A%2F%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%2F%2A%20String%20%2A%2F%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%2F%2A%20Comment%20%2A%2F%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%2F%2A%20OtherToken%20%2A%2F%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20AlertToken%20%2A%2F%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%2F%2A%20Function%20calls%20%2A%2F%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%2F%2A%20ErrorTok%20%2A%2F%0A%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div id="header">
<h1 class="title"><a href="https://github.com/wlandau/fbseqStudies"><code>fbseqStudies</code></a> package installation and usage guide</h1>
<h4 class="author"><em>Will Landau</em></h4>
<h4 class="date"><em>2015</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#purpose"><span class="toc-section-number">1</span> Purpose</a></li>
<li><a href="#system-requirements"><span class="toc-section-number">2</span> System requirements</a></li>
<li><a href="#installation"><span class="toc-section-number">3</span> Installation</a><ul>
<li><a href="#option-1-install-a-stable-release-recommended."><span class="toc-section-number">3.1</span> Option 1: install a stable release (recommended).</a></li>
<li><a href="#option-2-use-install_github-to-install-the-development-version."><span class="toc-section-number">3.2</span> Option 2: use <code>install_github</code> to install the development version.</a></li>
<li><a href="#option-3-build-the-development-version-from-the-source."><span class="toc-section-number">3.3</span> Option 3: build the development version from the source.</a></li>
</ul></li>
<li><a href="#run-a-scaled-down-version-first"><span class="toc-section-number">4</span> Run a scaled-down version first</a></li>
<li><a href="#replicate-the-studies"><span class="toc-section-number">5</span> Replicate the studies</a></li>
</ul>
</div>

<div id="purpose" class="section level1">
<h1><span class="header-section-number">1</span> Purpose</h1>
<p><a href="https://github.com/wlandau/fbseqStudies"><code>fbseqStudies</code></a> reproduces the results of academic studies of the methodology behind the <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a> and <a href="https://github.com/wlandau/fbseqCUDA"><code>fbseqCUDA</code></a> packages.</p>
</div>
<div id="system-requirements" class="section level1">
<h1><span class="header-section-number">2</span> System requirements</h1>
<ul>
<li>Mac OS X or Linux.</li>
<li>The required R version and R packages listed in the “Depends”, “Imports”, and “Suggests” fields of the “package’s <a href="https://github.com/wlandau/fbseqStudies/blob/master/DESCRIPTION">DESCRIPTION</a> file.</li>
<li>A CUDA-enabled NVIDIA graphics processing unit (GPU) with compute capability 2.0 or greater.</li>
<li>CUDA version 6.0 or greater. More information about CUDA is available through <a href="http://www.nvidia.com/object/cuda_home_new.html">NVIDIA</a>.</li>
</ul>
</div>
<div id="installation" class="section level1">
<h1><span class="header-section-number">3</span> Installation</h1>
<div id="option-1-install-a-stable-release-recommended." class="section level2">
<h2><span class="header-section-number">3.1</span> Option 1: install a stable release (recommended).</h2>
<p>Navigate to a <a href="https://github.com/wlandau/fbseqStudies/releases">list of stable releases</a> on the project’s <a href="https://github.com/wlandau/fbseqStudies">GitHub page</a>. Download the desired <code>tar.gz</code> bundle, then install it either with <code>install.packages(..., repos = NULL, type=&quot;source&quot;)</code> from within R <code>R CMD INSTALL</code> from the Unix/Linux command line.</p>
</div>
<div id="option-2-use-install_github-to-install-the-development-version." class="section level2">
<h2><span class="header-section-number">3.2</span> Option 2: use <code>install_github</code> to install the development version.</h2>
<p>For this option, you need the <code>devtools</code> package, available from CRAN or GitHub. Open R and run</p>
<pre><code>library(devtools)
install_github(&quot;wlandau/fbseqStudies&quot;)</code></pre>
</div>
<div id="option-3-build-the-development-version-from-the-source." class="section level2">
<h2><span class="header-section-number">3.3</span> Option 3: build the development version from the source.</h2>
<p>Open a command line program such as Terminal in Mac/Linux and enter the following commands.</p>
<pre><code>git clone git@github.com:wlandau/fbseqStudies.git
R CMD build fbseqStudies
R CMD INSTALL ...</code></pre>
<p>where <code>...</code> is replaced by the name of the tarball produced by <code>R CMD build</code>.</p>
</div>
</div>
<div id="run-a-scaled-down-version-first" class="section level1">
<h1><span class="header-section-number">4</span> Run a scaled-down version first</h1>
<p>Before seriously running a long job with one of the functions in the next section, do a trial run first. For a trial run, set the <code>&quot;fbseqStudies.scaledown&quot;</code> option to <code>TRUE</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(fbseqStudies)
<span class="kw">options</span>(<span class="st">&quot;fbseqStudies.scaledown&quot;</span> =<span class="st"> </span>T) <span class="co"># Scale down the computation.</span>
<span class="kw">paper_case</span>() <span class="co"># Replicate the results of the case study paper.</span></code></pre>
<p>Calling <code>options(&quot;fbseqStudies.scaledown&quot; = T)</code> selects the serial backend for <a href="https://github.com/wlandau/fbseq"><code>fbseq</code></a>, ensures that datasets are small in the numbers of genes, and configures the MCMCs to run for only a few iterations. That way, <code>paper_case()</code> will complete in a few minutes on your home computer, as opposed to several days on a machine with a CUDA-capable general-purpose graphics processing unit (GPU).</p>
</div>
<div id="replicate-the-studies" class="section level1">
<h1><span class="header-section-number">5</span> Replicate the studies</h1>
<p>Each function below reproduces the results of a paper.</p>
<table>
<tbody>
<tr class="odd">
<td align="left"><code>paper_computation()</code></td>
<td align="left">publication pending</td>
</tr>
<tr class="even">
<td align="left"><code>paper_case()</code></td>
<td align="left">publication pending</td>
</tr>
<tr class="odd">
<td align="left"><code>paper_priors()</code></td>
<td align="left">publication pending</td>
</tr>
</tbody>
</table>
<p>To run all 3 above functions in sequence, call the function <code>Landau_dissertation()</code>, which reproduces all the computation, figures, tables, etc. of the Statistics PhD dissertation of Will Landau (<a href="http://will-landau.com" class="uri">http://will-landau.com</a>, <a href="mailto:will.landau@gmail.com">will.landau@gmail.com</a>).</p>
<p>For further control, choose among the following. For even finer-grained control, see the manual and help files.</p>
<table>
<tbody>
<tr class="odd">
<td align="left"><code>progress()</code></td>
<td align="left">In the above 3 <code>paper_*</code> functions, the rate limiting step is to produce directories called <code>real_mcmc</code>, <code>coverage_mcmc</code>, etc., each with <code>.rds</code> files inside. Each <code>.rds</code> file contains a single (simulated or real) RNA-seq dataset and all the analyses of that dataset. To check the progress of the analyses, run <code>progress(&quot;coverage_mcmc&quot;)</code>, for example. Running <code>progress(&quot;coverage_mcmc&quot;)</code> lists all the methods used to analyze each dataset in the <code>coverage_mcmc</code> directory produced by <code>coverage_mcmc()</code> in <code>paper_case()</code>.</td>
</tr>
<tr class="even">
<td align="left"><code>real_mcmc()</code></td>
<td align="left">Run different versions of the model on the <span class="citation">Paschold et al. (2012)</span> dataset.</td>
</tr>
<tr class="odd">
<td align="left"><code>computation_mcmc()</code></td>
<td align="left">Duplicate sections of the <span class="citation">Paschold et al. (2012)</span> dataset to create datasets of varying size. Fit the default model to see how runtime scales with the number of genes and the number of libraries.</td>
</tr>
<tr class="even">
<td align="left"><code>coverage_mcmc()</code></td>
<td align="left">Simulate different datasets from the model to assess the ability to recapture parameters of interest in credible intervals calculated from the estimated full joint posterior distribution.</td>
</tr>
<tr class="odd">
<td align="left"><code>comparison_mcmc()</code></td>
<td align="left">Simulate datasets from multiple scenarios and fit multiple versions of the model to assess gene detection power and the calibration of posterior probabilites.</td>
</tr>
<tr class="even">
<td align="left"><code>serial_runs()</code></td>
<td align="left">Run various analyses, including a fully Bayesian one, of the <span class="citation">Paschold et al. (2012)</span> dataset in a version of the software that makes no use of GPU computing at all.</td>
</tr>
<tr class="odd">
<td align="left"><code>real_analyze()</code></td>
<td align="left">Extract figures, tables, and other results from the output of <code>real_mcmc()</code></td>
</tr>
<tr class="even">
<td align="left"><code>computation_analyze()</code></td>
<td align="left">Similar, but for <code>computation_mcmc()</code></td>
</tr>
<tr class="odd">
<td align="left"><code>coverage_analyze()</code></td>
<td align="left">Similar, but for <code>coverage_mcmc()</code></td>
</tr>
<tr class="even">
<td align="left"><code>comparison_analyze()</code></td>
<td align="left">Similar, but for <code>comparison_mcmc()</code></td>
</tr>
<tr class="odd">
<td align="left"><code>paper_case_figures()</code></td>
<td align="left">Once the <code>_mcmc()</code> and <code>_analyze()</code> functions are run in the current working directory, run this function to reproduce all the figures and tables of the computation paper (inside <code>paper_computation()</code>).</td>
</tr>
<tr class="even">
<td align="left"><code>paper_case_figures()</code></td>
<td align="left">Once the <code>_mcmc()</code> and <code>_analyze()</code> functions are run in the current working directory, run this function to reproduce all the figures and tables of the priors paper (inside <code>paper_priors()</code>).</td>
</tr>
</tbody>
</table>
<div class="references">
<h1 id="references" class="unnumbered">References</h1>
<p>Paschold, Anja, Yi Jia, Caroline Marcon, Steve Lund, Nick B Larson, Cheng-Ting Yeh, Stephan Ossowski, et al. 2012. “Complementation Contributes to Transcriptome Complexity in Maize (<em>Zea Mays L.</em>) Hybrids Relative to Their Inbred Parents.” <em>Genome Research</em> 22 (12). Cold Spring Harbor Lab: 2445–54.</p>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
